test/support/utils.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+require 'pty'
+require 'expect'
+require 'timeout'
+
+module DEBUGGER__
+  #
+  # Misc tools for the test suite
+  #
+  module TestUtils
+    def type(command)
+      @queue.push(command)
+    end
+
+    def assert_line_num(expected)
+      @queue.push(Proc.new { |result|
+        assert_equal(expected, result)
+      })
+    end
+
+    def debug_code(program, &block)
+      @queue = Queue.new
+      block.call
+      write_temp_file(program)
+      create_psuedo_terminal
+    end
+
+    def take_number(sentence)
+      sentence.match(/(.*):(.*)\r/)[2].to_i
+    end
+
+    def create_psuedo_terminal
+      PTY.spawn("ruby -r debug/run #{temp_file_path}") do |read, write, pid|
+        quitted = true
+        result = nil
+
+        while quitted
+          read.expect(/(.*)\n|\(rdbg\)/) do |sentence|
+            print sentence[0]
+            if sentence[0] == '(rdbg)'
+              cmd = @queue.pop
+              if cmd.is_a?(Proc)
+                cmd.call(result)
+                cmd = @queue.pop
+              end
+              write.puts(cmd)
+              quitted = false if cmd == 'quit'
```

Comment:
just curious about what `quitted` means here. I thought if `quitted == true`, the terminal would stop. but it looks like it's the opposite?

Reply:
Oh, that was my mistake. Thanks!